Dodanie potrzebnych bibliotek


In [120]:
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import confusion_matrix
from sklearn.naive_bayes import GaussianNB
import pandas as pd


Wczytanie danych z pliku `irys.csv`, podzielenie ich na zbiór treningowy oraz testowy a następnie wyodrębnienie z nich danych wejściowych oraz klas


In [121]:
df = pd.read_csv("iris.csv")

(train_set, test_set) = train_test_split(
    df.values, train_size=0.7, random_state=278804)

train_inputs = train_set[:, 0:4]
train_classes = train_set[:, 4]
test_inputs = test_set[:, 0:4]
test_classes = test_set[:, 4]


Przeprowadzenie klasyfikacji metodą k-najbliższych sąsiadów dla:

- 3 sąsiadów
- 5 sąsiadów
- 11 sąsiadów

oraz zapisanie ich dokładności oraz macierzy błędu


In [122]:
knn = {}

for neigbhours in [3, 5, 11]:
    tmp_knn = KNeighborsClassifier(n_neighbors=neigbhours)
    tmp_knn.fit(train_inputs, train_classes)

    score = tmp_knn.score(test_inputs, test_classes)
    prediction = tmp_knn.predict(test_inputs)
    matrix = confusion_matrix(test_classes, prediction)

    knn[neigbhours] = {"score": score,
                       "confusion_matrix": matrix}


Przeprowadzenie klasyfikacji klasyfikatorem Naive Bayes oraz zapisanie jej dokładności i macierzy błędu


In [123]:
nb = GaussianNB()
nb.fit(train_inputs, train_classes)

nb_score = nb.score(test_inputs, test_classes)
nb_prediction = nb.predict(test_inputs)
nb_matrix = confusion_matrix(test_classes, nb_prediction)


Wyświetlenie wyników i macierzy błędu dla wszystkich klasyfikatorów:

- 3-najbliższych sąsiadów


In [124]:
print(knn[3]['score'])
print(knn[3]['confusion_matrix'])


0.9555555555555556
[[14  0  0]
 [ 0 16  2]
 [ 0  0 13]]


- 5-najbliższych sąsiadów


In [125]:
print(knn[5]['score'])
print(knn[5]['confusion_matrix'])


0.9333333333333333
[[14  0  0]
 [ 0 15  3]
 [ 0  0 13]]


- 11-najbliższych sąsiadów


In [126]:
print(knn[11]['score'])
print(knn[11]['confusion_matrix'])


0.9555555555555556
[[14  0  0]
 [ 0 16  2]
 [ 0  0 13]]


- Naive Bayes


In [127]:
print(nb_score)
print(nb_matrix)


0.9777777777777777
[[14  0  0]
 [ 0 17  1]
 [ 0  0 13]]


Z porównania wszystkich 5 klasyfikatorów (drzewo decyzyjne, 3-najbliższych sąsiadów, 5-najbliższych sąsiadów, 11-najbliższych sąsiadów, Naive Bayes) pod względem dokładności wynika, że najlepszy z nich jest Naive Bayes, który osiągnął dokładność na poziomie 98%, najgorszy jest klasyfikator 5-najbliższych sąsiadów z wynikiem 93% dokładności, a wszystkie pozostałe uzyskały dokładność na poziomie 95%
